### DISCUSSION TITLE
Water97FluidProperties Problems

### DISCUSSION THREAD

**Discussion by YuhangNIU** [idaholab/moose#17377] [Q&A General] [answered]
Hi All,

I want to use **drho_dT** in the **rho_from_p_T()** function in **ADMaterial**, but when I was typing:
_"Real rho, drho_dp, drho_dT;
_fp.rho_from_p_T(_press[_qp],_temperature[_qp],rho,drho_dp,drho_dT)"_
the errors occurred:
_"no matching function for call to 'SinglePhaseFluidProperties::rho_from_p_T.'"
and"candidate expects 2 arguments, 5 provided."_

How can I use drho_dT?

Thanks,
Yuhang

**Comment by GiudGiud** [comment, marked as answer]
Hello

Since you are using an ADMaterial you may want to leverage AD to compute the derivatives.
@lee7632 has a recent fix for these properties and should be able to help

Guillaume

**Reply by YuhangNIU** [reply]
Yes. I use the fluid property of drho/dT in ADkernel, thus I need to build it in ADMaterial. I don't know how to use the drho_dT in Water97FluidProperties?

**Reply by cpgr** [reply]
Are `_press` and `_temperature` AD variables? If so, that won't work, you would need `_press[qp].value()` or similar. 

But if they are AD, then why not use `DualReal rho = rho_from_p_T(_press, _temperature)` and then `rho` has the derivative you want?

**Reply by YuhangNIU** [reply]
Thank you for your replies.
The **_press** and **_temprature** are AD variables, and **_press[qp].value()** still doesn't work.
If I use **DualReal rho = rho_from_p_T(_press, _temperature)**, the **rho** can be obtained, but I want to get **drho/dT** rather than **rho**.

**Reply by cpgr** [reply]
I'm pretty sure this would have worked
```
Real rho, drho_dp, drho_dT;
_fp.rho_from_p_T(_press[_qp].value(), _temperature[_qp].value, rho, drho_dp, drho_dT);
```
and if it doesn't, check for typos.

But the
```
DualReal rho = rho_from_p_T(_press[_qp],  _temperature[_qp]);
```
calculates exactly the same things. You can get the derivate wrt T using
```
Real drho_dT = rho.derivatives()[T_index];
```
where `T_index` is the index of the temperature variable.

**Reply by YuhangNIU** [reply]
Thank you for your patience!

For the first way:
The errors in Terminal are as shown: 
_Creating Unity /home/1/projects/zebra/build/unity_src/material_Unity.C
Compiling C++ (in opt mode) /home/1/projects/zebra/build/unity_src/material_Unity.C...
In file included from /home/1/projects/zebra/build/unity_src/material_Unity.C:5:0:
/home/1/projects/zebra/src/material/TwophaseADMaterial.C: In instantiation of ‘void TwophaseADMaterial<<anonymous> >::computeQpProperties() [with ComputeStage compute_stage = (ComputeStage)0u]’:
/home/1/projects/zebra/src/material/TwophaseADMaterial.C:663:1:   required from here
/home/1/projects/zebra/src/material/TwophaseADMaterial.C:255:3: error: request for member ‘value’ in ‘(&((TwophaseADMaterial<(ComputeStage)0u>*)this)->TwophaseADMaterial<(ComputeStage)0u>::_press)->MooseArray<T>::operator[]<double>(((TwophaseADMaterial<(ComputeStage)0u>*)this)->TwophaseADMaterial<(ComputeStage)0u>::<anonymous>.ADMaterial<(ComputeStage)0u>::<anonymous>.Material::_qp)’, which is of non-class type ‘const double’
   _fp.rho_from_p_T(_press[_qp].value(),_temp_liq[_qp].value(),rho,drho_dp,drho_dT);
   ^
/home/1/projects/zebra/src/material/TwophaseADMaterial.C:255:3: error: request for member ‘value’ in ‘(&((TwophaseADMaterial<(ComputeStage)0u>*)this)->TwophaseADMaterial<(ComputeStage)0u>::_temp_liq)->MooseArray<T>::operator[]<double>(((TwophaseADMaterial<(ComputeStage)0u>*)this)->TwophaseADMaterial<(ComputeStage)0u>::<anonymous>.ADMaterial<(ComputeStage)0u>::<anonymous>.Material::_qp)’, which is of non-class type ‘const double’
make: *** [/home/1/projects/zebra/build/unity_src/material_Unity.x86_64-pc-linux-gnu.opt.lo] Error 1_


For the second way:
How can I get the index of the temperature variable on the current point(_qp)?

What should I do?

Best wishes!

**Reply by cpgr** [reply]
I just checked the first way you tried and it definitely works (it's actually how the AD function is implemented). You must have a typo, or the variables aren't AD, or an old moose version, etc.

**Reply by cpgr** [reply]
Can you make your code available so we can take a look? It might make this easier.

**Reply by YuhangNIU** [reply]
Thank you for your reply,
I have solved this problem. Maybe the reason is that I have an old moose version. 
When I used **MetaPhysicL::raw_value(_press[_qp])** instead of **_press[_qp].value()**, it worked!

Thank you again for your patience！
Best wishes!
